yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,178 @@
+                from yt.utilities.on_demand_imports import _pyart
+
+                pyart = _pyart.pyart
+
+                radar = pyart.io.read_cfradial(filename)
+                self.grid_shape = grid_shape
+                self.grid_limits = grid_limits
+                grid = pyart.map.grid_from_radars(
+                    (radar,), grid_shape=self.grid_shape, grid_limits=self.grid_limits
+                )
+                mylog.warning(
+                    "Saving a cartesian grid for file %s at %s. Data will be loaded "
+                    "from the cartesian grid.",
+                    filename,
+                    storage_filename,
+                )
+                grid.write(storage_filename)
+            self._handle = xr.open_dataset(storage_filename)
+            filename = storage_filename
+        super().__init__(filename, dataset_type, units_override=units_override)
+        self.storage_filename = storage_filename
+        self.refine_by = 2  # refinement factor between a grid and its subgrid
+
+    def _set_code_unit_attributes(self):
+        length_unit = self._handle.variables["x"].attrs["units"]
+        self.length_unit = self.quan(1.0, length_unit)
+        self.mass_unit = self.quan(1.0, "kg")
+        self.time_unit = self.quan(1.0, "s")
+
+    def _parse_parameter_file(self):
+        self.parameters = {}
+
+        x, y, z = [self._handle.coords[d] for d in "xyz"]
+
+        self.origin_latitude = self._handle.origin_latitude[0]
+        self.origin_longitude = self._handle.origin_longitude[0]
+
+        self.domain_left_edge = np.array([x.min(), y.min(), z.min()])
+        self.domain_right_edge = np.array([x.max(), y.max(), z.max()])
+        self.dimensionality = 3
+        dims = [self._handle.dims[d] for d in "xyz"]
+        self.domain_dimensions = np.array(dims, dtype="int64")
+        self._periodicity = (False, False, False)
+        self.current_time = float(self._handle.time.values)
+
+        # Cosmological information set to zero (not in space).
+        self.cosmological_simulation = 0
+        self.current_redshift = 0.0
+        self.omega_lambda = 0.0
+        self.omega_matter = 0.0
+        self.hubble_constant = 0.0
+
+    @classmethod
+    def _is_valid(cls, filename, *args, **kwargs):
+        # This accepts a filename or a set of arguments and returns True or
+        # False depending on if the file is of the type requested.
+
+        conventions = ""
+        if xr._module:
+            try:
+                ds = xr.open_dataset(filename)
+            except (OSError, AttributeError, TypeError):
+                # catch all these to avoid errors when xarray cant handle a file
+                return False
+
+            if hasattr(ds, "attrs") and isinstance(ds.attrs, dict):
+                conventions += ds.attrs.get("Conventions", "")
+                conventions += ds.attrs.get("conventions", "")
+
+        return "CF/Radial" in conventions
```

Comment:
I tried to make this flatter, not crucial
```suggestion
        if not xr._module:
            return False
        try:
            ds = xr.open_dataset(filename)
        except (OSError, AttributeError, TypeError):
            # catch all these to avoid errors when xarray cant handle a file
            return False

        if hasattr(ds, "attrs") and isinstance(ds.attrs, dict):
            return "CF/Radial" in ds.attrs.get("Conventions", "") + ds.attrs.get("conventions", "")
        return False
```

Reply:
nice! will add it in after I see how the tests do...

Reply:
well I ended up adjusting `is_valid` a couple times -- didn't realize that `xr._module` would raise an `ImportError` if there is no `xarray`. The code now is fairly compact (and works!).